### DISCUSSION TITLE
How can we call the shortcodes in flynt ?

### DISCUSSION THREAD

**Discussion by rajpatel2435** [flyntwp/flynt#541] [Q&A] [answered]
Hey guys, is there anyone who has the complete documentation for flynt? As I started working on this theme for my new project. But I am not sure how can i start? Any advice would be highly appreciated!!

Thanks

**Comment by aaronmeder** [comment]
Hi @rajpatel2435 

Very nice to meet you, welcome on Planet Flynt 🥳

**Documentation**
There isn't a ton of official documentation for Flynt. I think this is also due to the fact that Flynt "does not do so much". Flynt mainly connects Wordpress, ACF, Timber, SCSS and web components in a convenient package. So if you're looking for solutions to challenges they are often more related  to these parts than Flynt itself.

For Flynt I would recommend reading through the [repo README](https://github.com/flyntwp/flynt) and the [Getting Started](https://flyntwp.com/the-beginners-guide-to-developing-a-custom-wordpress-theme-with-flynt/) page. 

If you're looking for a more high level info about what makes Flynt great, I've recorded a [video](https://youtu.be/8as8rVNcJuU) about that (the video shows Flynt v1 though, which was the same concepts, but some code details are different).

I do also want to record a more hands-on, rougher getting started video series where a simple site is created that shows a few nice tricks with Flynt...but it's hard to find the time for it at the moment 😬

**Shortcodes**
If you would like to render a shortcode in your Twig files, the easiest way is to use the ["shortcodes" filter]({{ post.custom_shortcode_field|shortcodes }}) like this:
```html
<div>{{ someHtml|shortcodes }}</div>
```

Does that help you to get on? 
Otherwise feel free to ask more specifically or post some code parts you're struggling with.

All the best from 🇨🇭!

Aaron

**Comment by rajpatel2435** [comment]
HI @aaronmeder 

Thank you for your prompt response.

let's say my shortcodes coming from plugin file. name of shortcodes is [my_shortcodes].
`{# index.twig inside your Flynt component #}


<flynt-component load:on="idle" load:on:media="(min-width: 1024px)" name="ShortcodeFront">
<div>
    <h2>Shortcode Values</h2>

    {{ function('do_shortcode', '[my_shortcode]') }}

    <ul>
        {% for value in shortcodeValues %}
            <li>{{ value }}</li>
        {% endfor %}
    </ul>
</div>
</flynt-component>

`

How can i use it here?

Thanks

**Reply by aaronmeder** [reply]
If your shortcode seems to return an array then you should be able to do this I think:

```html
{% set shortcodeValues = function('do_shortcode', '[my_shortcode]') %}

<ul>
    {% for value in shortcodeValues %}
        <li>{{ value }}</li>
    {% endfor %}
</ul>
```

If you would want to keep it a bit cleaner, then you could do the first part in your component's functions.php file:

```php
add_filter('Flynt/addComponentData?name=MyComponent', function (array $data): array {
    $data['shortcodeValues'] = do_shortcode('[my_shortcode]');
    return $data;
});
```

**Comment by rajpatel2435** [comment]
Hi @aaronmeder 

Here is the more clarity

index.twig 
![image](https://github.com/flyntwp/flynt/assets/81604941/652c9698-9972-4d78-81a9-1386fab2f046)
function.php
![image](https://github.com/flyntwp/flynt/assets/81604941/74bf83a3-4e40-4326-ab00-c83c1401a146)

result: empty screen
![image](https://github.com/flyntwp/flynt/assets/81604941/cd4649d4-7666-4c07-8bd6-649715750b54)

Any solution?

Thanks

**Reply by aaronmeder** [reply]
Yep I actually just realised when trying to test the same thing that a Wordpress shortcode will always return a text string, not an array...so you cannot use a for loop on it. 

Is using a shortcode for your purpose actually needed? What is it that you're trying to achieve?

**Comment by rajpatel2435** [comment]
Hi @aaronmeder

![image](https://github.com/flyntwp/flynt/assets/81604941/84eae04c-687f-4482-9224-5623f880ffa2)
actually we have one plugin and we are using the API to fetch the data and for output we are using the shortcodes to display that.

How can we achieve that?

**Reply by aaronmeder** [reply]
@rajpatel2435 Two ways how you could achieve that:

Either your shortcode returnes the finished, prepared html. Then you just render it directly in Twig like this: `{{ "[my_shortcode]"|shortcodes }}`

Or, if you want to create the html yourself using Twig, you would need to add a function to your plugin that just returns the needed data as an array. You can then call that function within your component's functions.php and loop through the data with Twig (how you would have originally wanted to do it as far as I understand):

**functions.php**
```php
add_filter('Flynt/addComponentData?name=MyComponentName', function (array $data): array {
    $data['items'] = function_that_returns_dashboard_data_as_array();
    return $data;
});
```

**index.twig**
```html
{% for item in items %}
  {{ item.date }} {{ item.time }}
{% endfor %}
```

Does that help?
Aaron

**Comment by rajpatel2435** [comment]
Hi @aaronmeder @timohubois 

it is not working even i am returning the prepared html.

here is the component structure

![image](https://github.com/flyntwp/flynt/assets/81604941/6cce6be4-3e21-4de6-8cd6-6e61771430cc)

functions.php
![image](https://github.com/flyntwp/flynt/assets/81604941/7e83aa76-a657-4ced-bf3a-b88f6a0e8f7e)

index.twig
![image](https://github.com/flyntwp/flynt/assets/81604941/d05f7e75-851e-4cbf-b92c-2e909f28b9d7)

output 
![image](https://github.com/flyntwp/flynt/assets/81604941/53a56d4a-64a3-49ca-9180-3abf30f42a2a)

could you please help me out?

![image](https://github.com/flyntwp/flynt/assets/81604941/9e952d36-cadc-4c82-bf4c-c62ecc714cd3)
could you also provide the solution for this?

Thanks


And the thing is working in my case
Whenever I am calling the shortocdes from the plugin file it is working but when i make shortcode inside function.php that is not working

**Reply by aaronmeder** [reply]
Hey @rajpatel2435 

Ah I think you're hitting a namespace problem...

When you add the "add_shortcode" function directly into your component's functions.php you need to do it like this:

Example: BlockWysiwyg/functions.php
```php

namespace Flynt\Components\BlockWysiwyg;

function my_shortcode_output()
{
    return "Some <strong>bold</strong> text returned by shortcode.";
}

add_shortcode('my_shortcode', 'Flynt\Components\BlockWysiwyg\my_shortcode_output');

```

then in your index.twig:

```html
{{ "[my_shortcode]"|shortcodes }}
```

Since each Flynt component uses it's own namespace you have to add that as a prefix to your shortcode callback function name (the second parameter of the [add_shortcode](https://developer.wordpress.org/reference/functions/add_shortcode/) function).

Hope this fixes it for you.

### No shortcode needed?

If you plan to add the function directly to your components functions.php then it wouldn't be necessary to go through a shortcode. You could instead do it like that:

```php

namespace Flynt\Components\BlockWysiwyg;

add_filter('Flynt/addComponentData?name=BlockWysiwyg', function ($data, $componentName) {
    $data['someData'] = my_custom_function_that_returns_html();

    return $data;
}, 10, 2);

function my_custom_function_that_returns_html()
{
    return "Some other <strong>bold</strong> text returned by shortcode.";
}
```

index.twig:
```html
{{ someData }}
```

In this case the function could not only return html but also an array that you could loop over in Twig, etc.

**Reply by rajpatel2435** [reply]
Hey @aaronmeder 

Thanks man it works.

As we are planning to build like user page, login, sign up means like social media plateform what do you think it is better to use flynt?

BTW Thank you so much for your help

**Reply by aaronmeder** [reply]
@rajpatel2435 glad to hear. Would be great if you can mark the discussion as answered then.

Sounds like an interesting project. I think in general it's good to use whatever you are most experienced with 😉 I haven't created many community sites like a social media service based on WordPress, but there might solutions to get started with that easier.

**Reply by rajpatel2435** [reply]
Hey @aaronmeder 

question for style.css whatever i am writing inside style.css it is not loading why?
![image](https://github.com/flyntwp/flynt/assets/81604941/ab3b5ac9-44c5-46ae-ab6a-8ea03a94bfe3)
![image](https://github.com/flyntwp/flynt/assets/81604941/ec0fe26d-8ba4-48e9-91a5-d53d344d1f3a)

**Reply by aaronmeder** [reply, marked as answer]
Hey @rajpatel2435 

The style.css file doesn't get loaded in the frontend.

You have to add your css to `assets/main.scss` and have a watcher (`npm run start`) running or build a bundle (`npm run build:production` for example). This gets compiled to a file in the dist that is loaded.

btw: It's best practice to use a new discussion per question, so other people can find a helpful answer to the same question later on more easily.
It would be great if you could mark this discussion as answered for the same reasons ✅ 

Wish you a good Easter weekend.
Aaron
